File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.util.LinkedList;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    static int[][] arr;
    static boolean[][] visited;
    static int[][] answer;
    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, -1, 0, 1};

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(
                    new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                if (st.hasMoreTokens()) {
                    str = st.nextToken("\n");
                } else {
                    str = br.readLine();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

        static class Node {
            int x, y, count;

            public Node(int x, int y, int count) {
                this.x = x;
                this.y = y;
                this.count = count;
            }
        }

        private static void bfs(int x, int y, int count) {
            int n = arr.length, m = arr[0].length;
            LinkedList<Node> q = new LinkedList<>();
            q.add(new Node(x, y, count));
            answer[x][y] = count;
            visited[x][y] = true;
            while (!q.isEmpty()) {
                Node node = q.removeFirst();
                int a = node.x, b = node.y, cur = node.count;
                for (int i = 0; i < 4; i++) {
                    int nx = a + dx[i], ny = b + dy[i];
                    if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                    if (visited[nx][ny]) continue;
                    if (arr[nx][ny] == 1 && !visited[nx][ny]) {
                        visited[nx][ny] = true;
                        answer[nx][ny] = cur + 1;
                        q.add(new Node(nx, ny, cur + 1));
                    }
                }
            }
        }

        public static void main(String[] args) {
            FastReader sc = new FastReader();
            int n = sc.nextInt(), m = sc.nextInt();
            Node start = null;
            arr = new int[n][m];
            answer = new int[n][m];
            visited = new boolean[n][m];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    arr[i][j] = sc.nextInt();
                    if (arr[i][j] == 2) {
                        start = new Node(i, j, 0);
                    }
                }
            }

            bfs(start.x, start.y, start.count);

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (i == start.x && j == start.y) {
                        System.out.print(answer[i][j] + " ");
                    } else if (arr[i][j] == 1 && answer[i][j] > 0) {
                        System.out.print(answer[i][j] + " ");
                    } else if (arr[i][j] == 0) {
                        System.out.print(0 + " ");
                    } else {
                        System.out.print(-1 + " ");
                    }
                }
                System.out.println();
            }
        }
}

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
from collections import deque
from sys import stdin
input = stdin.readline

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]


n, m = map(int, input().strip().split())
arr = []
visited = [[False] * m for _ in range(n)]
answer = [[0] * m for _ in range(n)]
start = (0, 0, 0)
for i in range(n):
    arr.append(list(map(int, input().strip().split())))
    for j in range(m):
        if arr[i][j] == 2:
            start = (i, j, 0)

def bfs(x, y, count):
    q = deque()
    q.append((x, y, count))
    answer[x][y] = count
    visited[x][y] = True
    while q:
        a, b, cur = q.popleft()
        for i in range(4):
            nx, ny = a + dx[i], b + dy[i]
            if 0 <= nx < n and 0 <= ny < m:
                if not visited[nx][ny] and arr[nx][ny] == 1:
                    visited[nx][ny] = True
                    answer[nx][ny] = cur + 1
                    q.append((nx, ny, cur + 1))

bfs(start[0], start[1], start[2])

for i in range(n):
    for j in range(m):
        if i == start[0] and j == start[1]:
            print(answer[i][j], end = " ")
        elif arr[i][j] == 1 and answer[i][j] != 0:
            print(answer[i][j], end = " ")
        elif arr[i][j] == 0:
            print(0, end = " ")
        else:
            print(-1, end = " ")
    print()



        